File: src/calib3d/calib3d_async.cpp
#include "calib3d_async.h"
#include "core/types.h"
#include "core/vec.hpp"
#include <opencv2/calib3d.hpp>

CvStatus *fisheye_undistortImage_Async(Mat distorted, Mat k, Mat d, CVD_OUT CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::fisheye::undistortImage(*distorted.ptr, dst, *k.ptr, *d.ptr);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *fisheye_undistortImageWithParams_Async(
    Mat distorted, Mat k, Mat d, Mat knew, Size size, CVD_OUT CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::fisheye::undistortImage(
      *distorted.ptr, dst, *k.ptr, *d.ptr, *knew.ptr, cv::Size(size.width, size.height)
  );
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *fisheye_undistortPoints_Async(
    Mat distorted, Mat k, Mat d, Mat R, Mat P, CVD_OUT CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::fisheye::undistortPoints(*distorted.ptr, dst, *k.ptr, *d.ptr, *R.ptr, *P.ptr);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *fisheye_estimateNewCameraMatrixForUndistortRectify_Async(
    Mat k,
    Mat d,
    Size imgSize,
    Mat r,
    double balance,
    Size newSize,
    double fovScale,
    CVD_OUT CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::fisheye::estimateNewCameraMatrixForUndistortRectify(
      *k.ptr,
      *d.ptr,
      cv::Size(imgSize.width, imgSize.height),
      *r.ptr,
      dst,
      balance,
      cv::Size(newSize.width, newSize.height),
      fovScale
  );
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

// calib3d
CvStatus *calibrateCamera_Async(
    VecVecPoint3f objectPoints,
    VecVecPoint2f imagePoints,
    Size imageSize,
    Mat cameraMatrix,
    Mat distCoeffs,
    int flag,
    TermCriteria criteria,
    CVD_OUT CvCallback_3 callback
) {
  BEGIN_WRAP
  auto tc = cv::TermCriteria(criteria.type, criteria.maxCount, criteria.epsilon);
  cv::Mat rvecs;
  cv::Mat tvecs;
  auto _objectPoints = vecvecpoint3f_c2cpp(objectPoints);
  auto _imagePoints = vecvecpoint2f_c2cpp(imagePoints);
  auto rval = cv::calibrateCamera(
      _objectPoints,
      _imagePoints,
      cv::Size(imageSize.width, imageSize.height),
      *cameraMatrix.ptr,
      *distCoeffs.ptr,
      rvecs,
      tvecs,
      flag,
      tc
  );
  callback(new double(rval), new Mat{new cv::Mat(rvecs)}, new Mat{new cv::Mat(tvecs)});
  END_WRAP
}
CvStatus *drawChessboardCorners_Async(
    Mat image, Size patternSize, bool patternWasFound, CVD_OUT CvCallback_0 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::drawChessboardCorners(
      *image.ptr, cv::Size(patternSize.width, patternSize.height), dst, patternWasFound
  );
  callback();
  END_WRAP
}
CvStatus *estimateAffinePartial2D_Async(VecPoint2f from, VecPoint2f to, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  auto _from = vecpoint2f_c2cpp(from);
  auto _to = vecpoint2f_c2cpp(to);
  cv::estimateAffinePartial2D(_from, _to, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}
CvStatus *estimateAffinePartial2DWithParams_Async(
    VecPoint2f from,
    VecPoint2f to,
    int method,
    double ransacReprojThreshold,
    size_t maxIters,
    double confidence,
    size_t refineIters,
    CVD_OUT CvCallback_2 callback
) {
  BEGIN_WRAP
  cv::Mat inliers;
  auto _from = vecpoint2f_c2cpp(from);
  auto _to = vecpoint2f_c2cpp(to);
  cv::Mat dst = cv::estimateAffinePartial2D(
      _from, _to, inliers, method, ransacReprojThreshold, maxIters, confidence, refineIters
  );
  callback(new Mat{new cv::Mat(dst)}, new Mat{new cv::Mat(inliers)});
  END_WRAP
}
CvStatus *estimateAffine2D_Async(VecPoint2f from, VecPoint2f to, CVD_OUT CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  auto _from = vecpoint2f_c2cpp(from);
  auto _to = vecpoint2f_c2cpp(to);
  cv::estimateAffine2D(_from, _to, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}
CvStatus *estimateAffine2DWithParams_Async(
    VecPoint2f from,
    VecPoint2f to,
    int method,
    double ransacReprojThreshold,
    size_t maxIters,
    double confidence,
    size_t refineIters,
    CVD_OUT CvCallback_2 callback
) {
  BEGIN_WRAP
  cv::Mat inliers;
  auto _from = vecpoint2f_c2cpp(from);
  auto _to = vecpoint2f_c2cpp(to);
  cv::Mat dst = cv::estimateAffine2D(
      _from, _to, inliers, method, ransacReprojThreshold, maxIters, confidence, refineIters
  );
  callback(new Mat{new cv::Mat(dst)}, new Mat{new cv::Mat(inliers)});
  END_WRAP
}
CvStatus *
findChessboardCorners_Async(Mat image, Size patternSize, int flags, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::Mat dst2;
  auto rval = cv::findChessboardCorners(
      *image.ptr, cv::Size(patternSize.width, patternSize.height), dst, flags
  );
  callback(new bool(rval), new Mat{new cv::Mat(dst)});
  END_WRAP
}
CvStatus *
findChessboardCornersSB_Async(Mat image, Size patternSize, int flags, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::Mat dst2;
  auto rval = cv::findChessboardCornersSB(
      *image.ptr, cv::Size(patternSize.width, patternSize.height), dst, flags
  );
  callback(new bool(rval), new Mat{new cv::Mat(dst)});
  END_WRAP
}
CvStatus *findChessboardCornersSBWithMeta_Async(
    Mat image, Size patternSize, int flags, CVD_OUT CvCallback_3 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::Mat dst2;
  auto rval = cv::findChessboardCornersSB(
      *image.ptr, cv::Size(patternSize.width, patternSize.height), dst, flags, dst2
  );
  callback(new bool(rval), new Mat{new cv::Mat(dst)}, new Mat{new cv::Mat(dst2)});
  END_WRAP
}
CvStatus *getOptimalNewCameraMatrix_Async(
    Mat cameraMatrix,
    Mat distCoeffs,
    Size size,
    double alpha,
    Size newImgSize,
    bool centerPrincipalPoint,
    CVD_OUT CvCallback_2 callback
) {
  BEGIN_WRAP
  cv::Rect dst;
  auto rval = cv::getOptimalNewCameraMatrix(
      *cameraMatrix.ptr,
      *distCoeffs.ptr,
      cv::Size(size.width, size.height),
      alpha,
      cv::Size(newImgSize.width, newImgSize.height),
      &dst,
      centerPrincipalPoint
  );
  callback(new Mat{new cv::Mat(rval)}, new Rect{dst.x, dst.y, dst.width, dst.height});
  END_WRAP
}
CvStatus *initUndistortRectifyMap_Async(
    Mat cameraMatrix,
    Mat distCoeffs,
    Mat r,
    Mat newCameraMatrix,
    Size size,
    int m1type,
    CVD_OUT CvCallback_2 callback
) {
  BEGIN_WRAP
  cv::Mat dst1;
  cv::Mat dst2;
  cv::initUndistortRectifyMap(
      *cameraMatrix.ptr,
      *distCoeffs.ptr,
      *r.ptr,
      *newCameraMatrix.ptr,
      cv::Size(size.width, size.height),
      m1type,
      dst1,
      dst2
  );
  callback(new Mat{new cv::Mat(dst1)}, new Mat{new cv::Mat(dst2)});
  END_WRAP
}

CvStatus *undistort_Async(
    Mat src, Mat cameraMatrix, Mat distCoeffs, Mat newCameraMatrix, CVD_OUT CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::undistort(*src.ptr, dst, *cameraMatrix.ptr, *distCoeffs.ptr, *newCameraMatrix.ptr);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}
CvStatus *undistortPoints_Async(
    Mat distorted, Mat k, Mat d, Mat r, Mat p, TermCriteria criteria, CVD_OUT CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::undistortPoints(
      *distorted.ptr,
      dst,
      *k.ptr,
      *d.ptr,
      *r.ptr,
      *p.ptr,
      cv::TermCriteria(criteria.type, criteria.maxCount, criteria.epsilon)
  );
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *FindHomography_Async(
    Mat src,
    Mat dst,
    int method,
    double ransacReprojThreshold,
    const int maxIters,
    const double confidence,
    CvCallback_2 callback
) {
  BEGIN_WRAP
  cv::Mat mask;
  cv::Mat out = cv::findHomography(
      *src.ptr, *dst.ptr, method, ransacReprojThreshold, mask, maxIters, confidence
  );
  callback(new Mat{new cv::Mat(out)}, new Mat{new cv::Mat(mask)});
  END_WRAP
}

File: lib/src/calib3d/calib3d_async.dart
library cv.calib3d;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/contours.dart';
import '../core/mat.dart';
import '../core/point.dart';
import '../core/rect.dart';
import '../core/size.dart';
import '../core/termcriteria.dart';
import '../g/calib3d.g.dart' as ccalib3d;
import '../g/constants.g.dart';

/// InitUndistortRectifyMap computes the joint undistortion and rectification transformation and represents the result in the form of maps for remap
///
/// For further details, please see:
/// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga7dfb72c9cf9780a347fbe3d1c47e5d5a
Future<(Mat, Mat)> initUndistortRectifyMapAsync(
  InputArray cameraMatrix,
  InputArray distCoeffs,
  InputArray R,
  InputArray newCameraMatrix,
  (int, int) size,
  int m1type,
) async =>
    cvRunAsync2<(Mat, Mat)>(
      (callback) => ccalib3d.initUndistortRectifyMap_Async(
        cameraMatrix.ref,
        distCoeffs.ref,
        R.ref,
        newCameraMatrix.ref,
        size.cvd.ref,
        m1type,
        callback,
      ),
      matCompleter2,
    );

/// GetOptimalNewCameraMatrixWithParams computes and returns the optimal new camera matrix based on the free scaling parameter.
///
/// For further details, please see:
/// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga7a6c4e032c97f03ba747966e6ad862b1
Future<(Mat rval, Rect validPixROI)> getOptimalNewCameraMatrixAsync(
  InputArray cameraMatrix,
  InputArray distCoeffs,
  (int, int) imageSize,
  double alpha, {
  (int, int) newImgSize = (0, 0),
  bool centerPrincipalPoint = false,
}) async =>
    cvRunAsync2(
      (callback) => ccalib3d.getOptimalNewCameraMatrix_Async(
        cameraMatrix.ref,
        distCoeffs.ref,
        imageSize.cvd.ref,
        alpha,
        newImgSize.cvd.ref,
        centerPrincipalPoint,
        callback,
      ),
      (c, p, p1) => c.complete((Mat.fromPointer(p.cast()), Rect.fromPointer(p1.cast()))),
    );

// CalibrateCamera finds the camera intrinsic and extrinsic parameters from several views of a calibration pattern.
//
// For further details, please see:
// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga3207604e4b1a1758aa66acb6ed5aa65d
Future<(double rmsErr, Mat cameraMatrix, Mat distCoeffs, Mat rvecs, Mat tvecs)> calibrateCameraAsync(
  Contours3f objectPoints,
  Contours2f imagePoints,
  (int, int) imageSize,
  InputOutputArray cameraMatrix,
  InputOutputArray distCoeffs, {
  Mat? rvecs,
  Mat? tvecs,
  int flags = 0,
  (int type, int count, double eps) criteria = (TERM_COUNT + TERM_EPS, 30, 1e-4),
}) async =>
    cvRunAsync3(
      (callback) => ccalib3d.calibrateCamera_Async(
        objectPoints.ref,
        imagePoints.ref,
        imageSize.cvd.ref,
        cameraMatrix.ref,
        distCoeffs.ref,
        flags,
        criteria.cvd.ref,
        callback,
      ),
      (c, p, p1, p2) {
        final rmsErr = p.cast<ffi.Double>().value;
        calloc.free(p);
        final rvecs = Mat.fromPointer(p1.cast());
        final tvecs = Mat.fromPointer(p2.cast());
        return c.complete((rmsErr, cameraMatrix, distCoeffs, rvecs, tvecs));
      },
    );

// Transforms an image to compensate for lens distortion.
// The function transforms an image to compensate radial and tangential lens distortion.
// The function is simply a combination of initUndistortRectifyMap (with unity R ) and remap (with bilinear interpolation). See the former function for details of the transformation being performed.
// Those pixels in the destination image, for which there is no correspondent pixels in the source image, are filled with zeros (black color).
// A particular subset of the source image that will be visible in the corrected image can be regulated by newCameraMatrix. You can use getOptimalNewCameraMatrix to compute the appropriate newCameraMatrix depending on your requirements.
// The camera matrix and the distortion parameters can be determined using calibrateCamera. If the resolution of images is different from the resolution used at the calibration stage, fx,fy,cx and cy need to be scaled accordingly, while the distortion coefficients remain the same.
Future<Mat> undistortAsync(
  InputArray src,
  InputArray cameraMatrix,
  InputArray distCoeffs, {
  InputArray? newCameraMatrix,
}) async =>
    cvRunAsync(
      (callback) => ccalib3d.undistort_Async(
        src.ref,
        cameraMatrix.ref,
        distCoeffs.ref,
        newCameraMatrix?.ref ?? Mat.empty().ref,
        callback,
      ),
      matCompleter,
    );

// UndistortPoints transforms points to compensate for lens distortion
//
// For further details, please see:
// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga55c716492470bfe86b0ee9bf3a1f0f7e
Future<Mat> undistortPointsAsync(
  InputArray src,
  InputArray cameraMatrix,
  InputArray distCoeffs, {
  InputArray? R,
  InputArray? P,
  (int type, int count, double eps) criteria = (TERM_COUNT + TERM_EPS, 30, 1e-4),
}) async =>
    cvRunAsync(
      (callback) => ccalib3d.undistortPoints_Async(
        src.ref,
        cameraMatrix.ref,
        distCoeffs.ref,
        R?.ref ?? Mat.empty().ref,
        P?.ref ?? Mat.empty().ref,
        criteria.cvd.ref,
        callback,
      ),
      matCompleter,
    );

// FindChessboardCorners finds the positions of internal corners of the chessboard.
//
// For further details, please see:
// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga93efa9b0aa890de240ca32b11253dd4a
Future<(bool success, Mat corners)> findChessboardCornersAsync(
  InputArray image,
  (int, int) patternSize, {
  int flags = CALIB_CB_ADAPTIVE_THRESH + CALIB_CB_NORMALIZE_IMAGE,
}) async =>
    cvRunAsync2(
      (callback) => ccalib3d.findChessboardCorners_Async(
        image.ref,
        patternSize.cvd.ref,
        flags,
        callback,
      ),
      (c, p, p1) {
        final rval = p.cast<ffi.Bool>().value;
        calloc.free(p);
        final corners = Mat.fromPointer(p1.cast());
        return c.complete((rval, corners));
      },
    );

// Finds the positions of internal corners of the chessboard using a sector based approach.
// https://docs.opencv.org/4.x/d9/d0c/group__calib3d.html#gadc5bcb05cb21cf1e50963df26986d7c9
Future<(bool, Mat corners)> findChessboardCornersSBAsync(
  InputArray image,
  (int, int) patternSize,
  int flags,
) async =>
    cvRunAsync2(
      (callback) => ccalib3d.findChessboardCornersSB_Async(
        image.ref,
        patternSize.cvd.ref,
        flags,
        callback,
      ),
      (c, p, p1) {
        final rval = p.cast<ffi.Bool>().value;
        calloc.free(p);
        final corners = Mat.fromPointer(p1.cast());
        return c.complete((rval, corners));
      },
    );

// Finds the positions of internal corners of the chessboard using a sector based approach.
// https://docs.opencv.org/4.x/d9/d0c/group__calib3d.html#gadc5bcb05cb21cf1e50963df26986d7c9
Future<(bool, Mat corners, Mat meta)> findChessboardCornersSBWithMetaAsync(
  InputArray image,
  (int, int) patternSize,
  int flags,
) async =>
    cvRunAsync3(
      (callback) =>
          ccalib3d.findChessboardCornersSBWithMeta_Async(image.ref, patternSize.cvd.ref, flags, callback),
      (c, p, p1, p2) {
        final rval = p.cast<ffi.Bool>().value;
        calloc.free(p);
        final corners = Mat.fromPointer(p1.cast());
        final meta = Mat.fromPointer(p2.cast());
        return c.complete((rval, corners, meta));
      },
    );

// DrawChessboardCorners renders the detected chessboard corners.
//
// For further details, please see:
// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#ga6a10b0bb120c4907e5eabbcd22319022
Future<Mat> drawChessboardCornersAsync(
  InputOutputArray image,
  (int, int) patternSize,
  InputArray corners,
  bool patternWasFound,
) async =>
    cvRunAsync0<Mat>(
      (callback) =>
          ccalib3d.drawChessboardCorners_Async(image.ref, patternSize.cvd.ref, patternWasFound, callback),
      (c) => c.complete(image),
    );

// EstimateAffinePartial2D computes an optimal limited affine transformation
// with 4 degrees of freedom between two 2D point sets.
//
// For further details, please see:
// https://docs.opencv.org/master/d9/d0c/group__calib3d.html#gad767faff73e9cbd8b9d92b955b50062d
Future<(Mat, Mat inliers)> estimateAffinePartial2DAsync(
  VecPoint2f from,
  VecPoint2f to, {
  int method = RANSAC,
  double ransacReprojThreshold = 3,
  int maxIters = 2000,
  double confidence = 0.99,
  int refineIters = 10,
}) async =>
    cvRunAsync2(
      (callback) => ccalib3d.estimateAffinePartial2DWithParams_Async(
        from.ref,
        to.ref,
        method,
        ransacReprojThreshold,
        maxIters,
        confidence,
        refineIters,
        callback,
      ),
      matCompleter2,
    );

// EstimateAffine2D Computes an optimal affine transformation between two 2D point sets.
//
// For further details, please see:
// https://docs.opencv.org/4.0.0/d9/d0c/group__calib3d.html#ga27865b1d26bac9ce91efaee83e94d4dd
Future<(Mat, Mat inliers)> estimateAffine2DAsync(
  VecPoint2f from,
  VecPoint2f to, {
  int method = RANSAC,
  double ransacReprojThreshold = 3,
  int maxIters = 2000,
  double confidence = 0.99,
  int refineIters = 10,
}) async =>
    cvRunAsync2(
      (callback) => ccalib3d.estimateAffine2DWithParams_Async(
        from.ref,
        to.ref,
        method,
        ransacReprojThreshold,
        maxIters,
        confidence,
        refineIters,
        callback,
      ),
      matCompleter2,
    );
